atomic_reactor/plugins/pre_change_from_in_df.py

```diff
@@ -44,19 +44,29 @@ def __init__(self, tasker, workflow):
 
     def run(self):
         builder = self.workflow.builder
-        dfp = DockerfileParser(builder.df_path)
+        dfp = df_parser(builder.df_path)
 
+        # df_base is the raw image string from DockerfileParser
+        # build_base_image is an ImageName of a raw image string
+        # parent_images is a dict of raw image strings, indexed by ImageName
+        # the naming scheme is clunky, but we need to keep track of raw strings,
+        # ImageNames (which may have a default tag), and strings derived from ImageNames
         df_base = dfp.baseimage
-        build_base = builder.base_image.to_str()
+        df_base_image = ImageName.parse(df_base)
+        df_base_str = df_base_image.to_str()
+        build_base_image = builder.base_image
+        build_base_str = builder.base_image.to_str()
 
         # do some sanity checks to defend against bugs and rogue plugins
-
-        if build_base != builder.parent_images[df_base]:
+        # builder.base_image may have a default tag of latest, and builder.parent_images[]
+        # won't have a default tag. So convert builder.parent_images[] to an ImageName
+        # so we're comparing apples to apples.
+        if build_base_str != str(ImageName.parse(builder.parent_images[df_base_str])):
```

Comment:
What are the changes that we're gonna mess this up again in the future? 

Wouldn't it make it much simpler if keys in parent_image were simply ImageName objects? We could even make it a custom dict to enforce that.

Reply:
@sosiouxme this is your design element.

Reply:
I think with this behavior of ImageName modifying the image it gets, it is too confusing to try to keep straight which `foo_image` is a string versus an ImageName versus a string representation of one for purposes of comparison. I would suggest that anywhere we start thinking of something as an image name, we parse it into an ImageName immediately and store and compare that way.

So `parent_images` becomes a dict of ImageName => ImageName. `original_base_name` becomes an ImageName. `set_base_name` should probably accept either an ImageName or a string that is parsed into an ImageName like now. There are only a few places where it will need to become a string again -- I think just the metadata plugins and of course when writing to DF.

Does that sound good?

Reply:
Works for me. I'll go implement lucarva's suggestion.

Reply:
That was fun. Fixed I think.